### ISSUE TITLE
Erroneous validation additional properties (okay v5.0.10 but fails v5.0.11)

### ISSUE THREAD

**Issue by dcrossleyau** [aml-org/amf#1551]
* **I'm submitting a ...**
  - [x] bug report
  - [ ] feature request
  - [ ] question
  - [ ] other

* **Please tell us about your environment:**

  - AMF Version: v5.0.11
  - AMF Distribution: amf-client-js 
  - Operating System: macOS Catalina with node.js 14.20.0

---

Our project has many separate modules with RAML or OpenAPI definitions. We use amf-client-js to validate. Each time that we upgrade that dependency, we process all 350+ files. With v5.0.10 all API descriptions do pass validation, but with v5.0.11 there is one RAML-1.0 description that is now failing. There is a strange validation report.

The validation report seems to be saying that certain objects "should NOT have additional properties" (which is correct) but then goes on to say that other properties are required. However those properties are not part of that schema.

The sources are at https://github.com/folio-org/mod-patron/tree/master/ramls \
Note that there is a git submodule of shared resources, so need a recursive git clone. \
Also note that there were no recent git source changes.

Doing independent validation of the schema and its example shows that all is well:

```
z-schema --pedanticCheck account.json examples/account.json
schema validation passed
json #1 validation passed
```

This is the relevant portion of the validation report from amf-client-js v5.0.11 ...

```
---- validationResult:
ModelId: file://./ramls/patron.raml
Profile: RAML 1.0
Conforms: false
Number of results: 1

Level: Violation

- Constraint: http://a.ml/vocabularies/amf/validation#example-validation-error
  Message: charges[0].chargeAmount should NOT have additional properties
charges[0].chargeAmount should have required property 'charges'
charges[0].chargeAmount should have required property 'holds'
charges[0].chargeAmount should have required property 'loans'
charges[0].chargeAmount should have required property 'totalCharges'
charges[0].chargeAmount should have required property 'totalChargesCount'
charges[0].chargeAmount should have required property 'totalHolds'
charges[0].chargeAmount should have required property 'totalLoans'
charges[0].item should NOT have additional properties
charges[0].item should have required property 'charges'
charges[0].item should have required property 'holds'
charges[0].item should have required property 'loans'
charges[0].item should have required property 'totalCharges'
charges[0].item should have required property 'totalChargesCount'
charges[0].item should have required property 'totalHolds'
charges[0].item should have required property 'totalLoans'
holds[0].item should NOT have additional properties
holds[0].item should have required property 'charges'
holds[0].item should have required property 'holds'
holds[0].item should have required property 'loans'
holds[0].item should have required property 'totalCharges'
holds[0].item should have required property 'totalChargesCount'
holds[0].item should have required property 'totalHolds'
holds[0].item should have required property 'totalLoans'
loans[0].item should NOT have additional properties
loans[0].item should have required property 'charges'
loans[0].item should have required property 'holds'
loans[0].item should have required property 'loans'
loans[0].item should have required property 'totalCharges'
loans[0].item should have required property 'totalChargesCount'
loans[0].item should have required property 'totalHolds'
loans[0].item should have required property 'totalLoans'

  Severity: Violation
  Target: file://./ramls/patron.raml#/web-api/endpoint/%2Fpatron%2Faccount%2F%7Bid%7D/supportedOperation/get/returns/resp/200/payload/application%2Fjson/shape/schema/examples/example/default-example
  Property: file://./ramls/patron.raml#/web-api/endpoint/%2Fpatron%2Faccount%2F%7Bid%7D/supportedOperation/get/returns/resp/200/payload/application%2Fjson/shape/schema/examples/example/default-example
  Range: [(1,0)-(56,1)]
  Location: file://./ramls/examples/account.json
```

**Comment by looseale** [comment]
Hi @dcrossleyau.
I cannot reproduce the error of `additionalProperties`, but yes the other errors. I don't know how you are calling AMF to get those errors (in my test I only parse and validate), you should provide a minimal AMF code as is requested in the issue template.
Anyway, the other errors are also incorrect and I was able to reproduce them. I reported an internal issue W-11631501 with this (this is only for our tracking).

**Comment by looseale** [comment]
If you are forced to bump the AMF version for some reason, I saw that moving the schemas to a subfolder (not putting them in the same folder as the API, but creating a `schemas` folder for example) and changing the reference from the RAML (adding the `schemas/` in the `types` of the RAML) is a workaround for the issue.

**Comment by dcrossleyau** [comment]
Thanks. Okay, i will return.

**Comment by dcrossleyau** [comment]
We do deliberately keep up to verify and deploy each new AMF version. This particular 5.0.11 release was important, as it dealt with JSON Schema.

We have a large OSS project with ~100 separate repositories. Most of them do have the schema files at the same directory level as the parent RAML descriptions. Some have a mixture, with some schema in subdirectories.

As said in the initial report, all [other](https://dev.folio.org/reference/api/) repositories are okay, just this one has this strange behaviour, but okay with the previous release.

**Comment by dcrossleyau** [comment]
This is the reduced script used to parse and validate. \
Do: `node validate.js ramls/patron.raml`

```javascript
const amf = require('amf-client-js');
const inputFn = process.argv[2];
const client = amf.RAMLConfiguration.RAML10().baseUnitClient();
async function main() {
  const parsingResult = await client.parseDocument(`file://${inputFn}`);
  const validationResult = await client.validate(parsingResult.baseUnit);
  console.log('---- parsingResult:');
  console.log(parsingResult.toString());
  console.log('---- validationResult:');
  console.log(validationResult.toString());
  if (!parsingResult.conforms || !validationResult.conforms) {
    process.exitCode = 1;
  }
}
main();
```

**Comment by dcrossleyau** [comment]
The package [test-amf-1511.tar.gz](https://github.com/aml-org/amf/files/9403257/test-amf-1511.tar.gz) contains the reduced RAML and just the relevant schema files. The RAML is reduced to one endpoint. The "types" declaration section is retained. Those declared schema are referenced from the parent JSON Schema "account.json". Those "types" are also used for other endpoints in the original RAML description.

```
#%RAML 1.0
title: Patron Services (reduced for bug report)

types:
  account: !include account.json
  hold: !include hold.json
  loan: !include loan.json
  charge: !include charge.json
  money: !include money.json

/patron:
  /account:
    get:
      responses:
        200:
          body:
            application/json:
              type: account
              example: !include examples/account.json
```

That RAML is okay using v5.0.10

With v5.0.11 this result:

```
---- parsingResult:
Model: file://ramls/patron.raml
Conforms: true
Number of results: 0

---- validationResult:
ModelId: file://ramls/patron.raml
Profile: RAML 1.0
Conforms: false
Number of results: 1

Level: Violation

- Constraint: http://a.ml/vocabularies/amf/validation#example-validation-error
  Message: charges[0].chargeAmount should NOT have additional properties
charges[0].chargeAmount should have required property 'charges'
charges[0].chargeAmount should have required property 'holds'
charges[0].chargeAmount should have required property 'loans'
charges[0].chargeAmount should have required property 'totalCharges'
charges[0].chargeAmount should have required property 'totalChargesCount'
charges[0].chargeAmount should have required property 'totalHolds'
charges[0].chargeAmount should have required property 'totalLoans'
holds[0].item should NOT have additional properties
holds[0].item should have required property 'charges'
holds[0].item should have required property 'holds'
holds[0].item should have required property 'loans'
holds[0].item should have required property 'totalCharges'
holds[0].item should have required property 'totalChargesCount'
holds[0].item should have required property 'totalHolds'
holds[0].item should have required property 'totalLoans'
loans[0].item should NOT have additional properties
loans[0].item should have required property 'charges'
loans[0].item should have required property 'holds'
loans[0].item should have required property 'loans'
loans[0].item should have required property 'totalCharges'
loans[0].item should have required property 'totalChargesCount'
loans[0].item should have required property 'totalHolds'
loans[0].item should have required property 'totalLoans'

  Severity: Violation
  Target: file://ramls/patron.raml#/web-api/endpoint/%2Fpatron%2Faccount/supportedOperation/get/returns/resp/200/payload/application%2Fjson/shape/schema/examples/example/default-example
  Property: file://ramls/patron.raml#/web-api/endpoint/%2Fpatron%2Faccount/supportedOperation/get/returns/resp/200/payload/application%2Fjson/shape/schema/examples/example/default-example
  Range: [(1,0)-(56,1)]
  Location: file://ramls/examples/account.json
```

Now comment-out the declaration of the "type" of "charge". This result:

```
---- parsingResult:
Model: file://ramls/patron.raml
Conforms: true
Number of results: 0

---- validationResult:
ModelId: file://ramls/patron.raml
Profile: RAML 1.0
Conforms: false
Number of results: 1

Level: Violation

- Constraint: http://a.ml/vocabularies/amf/validation#example-validation-error
  Message: holds[0].item should NOT have additional properties
holds[0].item should have required property 'charges'
holds[0].item should have required property 'holds'
holds[0].item should have required property 'loans'
holds[0].item should have required property 'totalCharges'
holds[0].item should have required property 'totalChargesCount'
holds[0].item should have required property 'totalHolds'
holds[0].item should have required property 'totalLoans'
loans[0].item should NOT have additional properties
loans[0].item should have required property 'charges'
loans[0].item should have required property 'holds'
loans[0].item should have required property 'loans'
loans[0].item should have required property 'totalCharges'
loans[0].item should have required property 'totalChargesCount'
loans[0].item should have required property 'totalHolds'
loans[0].item should have required property 'totalLoans'

  Severity: Violation
  Target: file://ramls/patron.raml#/web-api/endpoint/%2Fpatron%2Faccount/supportedOperation/get/returns/resp/200/payload/application%2Fjson/shape/schema/examples/example/default-example
  Property: file://ramls/patron.raml#/web-api/endpoint/%2Fpatron%2Faccount/supportedOperation/get/returns/resp/200/payload/application%2Fjson/shape/schema/examples/example/default-example
  Range: [(1,0)-(56,1)]
  Location: file://ramls/examples/account.json
```

Similarly also comment-out the declaration of the "type" of "loans" and "holds". This result:

```
---- parsingResult:
Model: file://ramls/patron.raml
Conforms: true
Number of results: 0

---- validationResult:
ModelId: file://ramls/patron.raml
Profile: RAML 1.0
Conforms: true
Number of results: 0
```

So removing those type declarations forces it to validate. However those types are required by the actual RAML so cannot be removed.

**Comment by dcrossleyau** [comment]
Same with the recent v5.0.12

**Comment by looseale** [comment]
Hi @dcrossleyau 
The issue is fixed with commit 7e48312ca1e80ed370ecbf74c083e2059ee8cdae and it will be released in the next AMF version (scheduled for October 19)

**Comment by dcrossleyau** [comment]
Thanks.
